### PR TITLE
Pseudo simulations, interactive mode

### PR DESCRIPTION
**Pull request by akaija**
Created methods for pseudo-simulated data, using analytical equations for gas loading, surface area, and void fraction as functions of number density, average sigma-value, and/or average epsilon-value.

Created methods of an _interactive_ mode providing increased user output and pauses for user to confirm mutation strength calculations, retest results, and even use manual parent selection.

### CODE REVIEW

File: htsohm/db/material.py [hunk 1]
```diff
@@ -1,14 +1,17 @@
-
+import os
 import sys
 import uuid
 
 from sqlalchemy import Column, ForeignKey, Integer, String, Float, Boolean
 from sqlalchemy.sql import text
 from sqlalchemy.orm import relationship
 
+#import htsohm
```
Comment (paulboone) [inline htsohm/db/material.py]: delete?

File: htsohm/db/material.py [hunk 2]
```diff
@@ -289,3 +292,27 @@ def calculate_retest_result(self, tolerance):
...
         return not retest_failed
+
+    def artificial_void_fraction(self):
+#        htsohm_dir = os.path.dirname(os.path.dirname(htsohm.__file__))
+#        config_file = os.path.join(htsohm_dir, self.run_id, 'config.yaml')
+#        config = load_config_file(config_file)
+
+        return pseudo_void_fraction(
+                config, self.structure.number_density(),
+                self.structure.average_sigma())
```
Comment (paulboone) [inline htsohm/db/material.py]: config not defined? does this run?

File: htsohm/db/material.py [hunk 3]
```diff
@@ -289,3 +292,27 @@ def calculate_retest_result(self, tolerance):
         )
 
         return not retest_failed
+
+    def artificial_void_fraction(self):
+#        htsohm_dir = os.path.dirname(os.path.dirname(htsohm.__file__))
+#        config_file = os.path.join(htsohm_dir, self.run_id, 'config.yaml')
+#        config = load_config_file(config_file)
```
Comment (paulboone) [inline htsohm/db/material.py]: delete

File: htsohm/db/material.py [hunk 4]
```diff
@@ -289,3 +292,27 @@ def calculate_retest_result(self, tolerance):
...
+
+    def artificial_surface_area(self):
+        return pseudo_surface_area(
+                self.artificial_void_fraction(),
+                self.structure.average_sigma())
+
+    def artificial_gas_adsorption(self):
+#        htsohm_dir = os.path.dirname(os.path.dirname(htsohm.__file__))
+#        config_file = os.path.join(htsohm_dir, self.run_id, 'config.yaml')
+#        config = load_config_file(config_file)
```
Comment (paulboone) [inline htsohm/db/material.py]: delete

File: htsohm/db/material.py [hunk 5]
```diff
@@ -289,3 +292,27 @@ def calculate_retest_result(self, tolerance):
...
+        
+        return pseudo_gas_adsorption(
+                config, self.artificial_void_fraction(),
+                self.artificial_surface_area(),
+                self.structure.average_epsilon())
```
Comment (paulboone) [inline htsohm/db/material.py]: config not defined? does this run?

File: htsohm/db/structure.py [hunk 6]
```diff
@@ -57,3 +57,18 @@ def clone(self):
     def volume(self):
         """   """
         return self.lattice_constant_a * self.lattice_constant_b * self.lattice_constant_c
+
```
Comment (paulboone) [inline htsohm/db/structure.py]: Need comments for what these are doing... these are just for the artificial data, right?

File: htsohm/htsohm.py [hunk 7]
```diff
@@ -148,55 +148,100 @@ def run_all_simulations(material):
...
+        else:
+            results = {}
+            results['sa_volumetric_surface_area'] = material.artificial_surface_area()
+            material.update_from_dict(results)
+            material.surface_area_bin = calc_bin(
+                material.sa_volumetric_surface_area,
+                *config['surface_area']['limits'],
+                config['number_of_convergence_bins']
+            )
+
```
Comment (paulboone) [inline htsohm/htsohm.py]: I see what you're doing here, and it should work this way.

We ought to be able to replace this whole block of code though with just a config file that lists the tests to run, and then run through them. Let's discuss in person.

File: htsohm/pseudo_material.py [hunk 8]
```diff
@@ -104,3 +108,36 @@ def volume(self):
 
     def number_density(self):
         return len(self.atom_sites) / self.volume()
+
```
Comment (paulboone) [inline htsohm/pseudo_material.py]: How are these different from the ones defined above? Do we really need two sets?

File: htsohm/pseudo_simulation.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import os
+
+#import htsohm
+#from htsohm.files import load_config_file
+
```
Comment (paulboone) [inline htsohm/pseudo_simulation.py]: clean up

File: htsohm/htsohm.py [hunk 10]
```diff
@@ -132,6 +132,83 @@ def select_parent(run_id, max_generation, generation_limit):
...
+    potential_parents = [i[0] for i in parent_query]
+
+    print('Potential parents :')
+    for i in potential_parents:
+        print('\t{}'.format(i))
+
+    parent_uuid = input('Please select a parent UUID :\t')
+
+    return session.query(Material.id).filter(Material.run_id == run_id, Material.uuid == parent_uuid).one()[0]
+
```
Comment (paulboone) [inline htsohm/htsohm.py]: All this code seems mostly duplicated from select_parent. We have to find a way for this code to be DRY, or it will be a problem going forward

File: htsohm/htsohm.py [hunk 11]
```diff
@@ -463,8 +540,13 @@ def worker_run_loop(run_id):
...
                 print("selecting a parent / running retests on parent / mutating / simulating")
-                parent_id = select_parent(run_id, max_generation=(gen - 1),
-                                                  generation_limit=config['children_per_generation'])
+                
+                if config['selection_mode'] == 'manual':
+                     parent_id = select_parent_interactive(run_id, max_generation=(gen - 1),
+                                                      generation_limit=config['children_per_generation'])
+                else:
+                    parent_id = select_parent(run_id, max_generation=(gen - 1),
+                                                      generation_limit=config['children_per_generation'])
```
Comment (paulboone) [inline htsohm/htsohm.py]: is there some sort of an error thrown on startup if this is being run as a worker on a node but the config is set to interactive?

File: htsohm/material_files.py [hunk 12]
```diff
@@ -187,6 +213,13 @@ def mutate_material(parent_material, mutation_strength, generation):
         min(number_of_atoms, len(parent_material.structure.atom_sites)),
         replace = False).tolist()
 
+    if config['interactive_mode'] == 'on':
+        p_x, p_y, p_z = [], [], []
+        for atom_site in child_material.structure.atom_sites:
+            p_x.append(atom_site.x_frac)
+            p_y.append(atom_site.y_frac)
+            p_z.append(atom_site.z_frac)
+
```
Comment (paulboone) [inline htsohm/material_files.py:224]: why are p_x, p_y, p_z calculated here if they are only used below?

File: htsohm/htsohm.py [hunk 13]
```diff
@@ -430,14 +430,20 @@ def calculate_percent_children_in_bin(run_id, generation, bin_coordinate):
...
+
     return len([ r for r in rows if r.in_bin ]) / len(rows)
 
 def calculate_mutation_strength(run_id, generation, mutation_strength_bin):
     """Query mutation_strength for bin and adjust as necessary.
 
     Args:
         run_id (str): identification string for run.
-        generation (int): iteration in bin-mutate-simulate routine.
+        generation (int): iteration in bin-mutate-/workersimulate routine.
```
Comment (paulboone) [inline htsohm/htsohm.py]: typo?

File: htsohm/htsohm.py [hunk 14]
```diff
@@ -458,14 +464,35 @@ def calculate_mutation_strength(run_id, generation, mutation_strength_bin):
...
+                while True:
+                    mutation_strength.strength = input('Please Enter new mutation strength :\t')
+                    proceed = input('Are you sure you want to assign {} to bin {} ? (y/n) :\t'.format(
+                        mutation_strength.strength, mutation_strength_bin))
+                    if proceed in ['n', 'N']:
+                        continue
+                    elif proceed in ['y', 'Y']:
+                        break
+                    else:
+                        print('Please Enter y/n.')
```
Comment (paulboone) [inline htsohm/htsohm.py]: You have this pattern in multiple places. Make it into a function and reuse?

File: htsohm/htsohm.py [hunk 15]
```diff
@@ -474,7 +501,7 @@ def calculate_mutation_strength(run_id, generation, mutation_strength_bin):
             session.commit()
         except (FlushError, IntegrityError) as e:
             print("Somebody beat us to saving a row with this generation. That's ok!")
-            session.rollback()
+            Session.rollback()
```
Comment (paulboone) [inline htsohm/htsohm.py]: Why did you change this? I don't think this is right... did you try it in the console?

File: htsohm/htsohm.py [hunk 16]
```diff
@@ -590,40 +617,60 @@ def worker_run_loop(run_id):
...
+                            all_accessed_bins = [ [e[0], e[1], e[2]]
+                                    for e in all_accessed_bin_tuples]
+                            print_block('ANNEALING WITH MUTATION STRENGTH :\t{}' \
+                                    .format(config['initial_mutation_strength']))
+                            for some_bin in all_accessed_bins:
+                                print('Annealing bin :\t{}'.format(some_bin))
+                                mutation_strength = MutationStrength(run_id, gen + 1, *some_bin)
+                                mutation_strength.strength = config['annealing_strength']
+                                session.add(mutation_strength)
+
```
Comment (paulboone) [inline htsohm/htsohm.py]: The worker_run_loop method is a beast and needs to be broken up into smaller submethods...

File: htsohm/htsohm.py [hunk 17]
```diff
@@ -376,13 +382,91 @@ def retest(m_orig, retests, tolerance):
...
+                        break
+                    else:
+                        print('Please enter (y/n).')
+                double_check = input('Are you sure? (y/n) :\t')
+                if double_check in ['y', 'Y']:
+                    break
+                elif double_check in ['n', 'N']:
+                    continue
+                else:
+                    print('Please enter (y/n).')
```
Comment (paulboone) [inline htsohm/htsohm.py]: same pattern again x2

File: htsohm/htsohm.py [hunk 18]
```diff
@@ -700,61 +770,11 @@ def worker_run_loop(run_id):
...
-                        for some_bin in ms_bins:
-                            print('\nCalculating strength for bin :\t{}'.format(some_bin))
-                            print('Bin {} / {}'.format(counter, len(ms_bins)))
-                            calculate_mutation_strength(run_id, gen + 1, parent_bin)
-                            input('Press Enter to continue :')
+            # make sure worker just finished thev last material in its generation
+            if material.generation_index == config['children_per_generation'] - 1 and gen > 0:
+                if config['mutation_strength_method'] != 'flat':
+                    calculate_all_mutation_strengths(config, gen)
+
```
Comment (paulboone) [inline htsohm/htsohm.py]: A great start on making worker_run_loop simpler!

It's hard to tell from the diff because of Python's whitespace formatting. Was there more changed here besides refactoring it into a separate method?

**Review by paulboone — CHANGES_REQUESTED** [review]
Generally looks good! Some small stuff and some stuff we should discuss in person.

File: htsohm/htsohm.py (hunk 11 above)
Comment (akaija) [reply htsohm/htsohm.py]: Nope!

File: htsohm/db/material.py (hunk 2 above)
Comment (akaija) [reply htsohm/db/material.py]: it's a global variable -- something to discuss? the global config makes things less portable.

File: htsohm/db/material.py (hunk 5 above)
Comment (akaija) [reply htsohm/db/material.py]: global variable. see previous comment.

File: htsohm/htsohm.py (hunk 7 above)
Comment (akaija) [reply htsohm/htsohm.py]: Ping! Let's talk about this.

File: htsohm/htsohm.py (hunk 10 above)
Comment (akaija) [reply htsohm/htsohm.py]: Done--mostly!

File: htsohm/htsohm.py [hunk 19]
```diff
@@ -290,8 +484,26 @@ def calculate_percent_children_in_bin(run_id, generation, bin_coordinate):
...
+        value = input(message)
+        proceed = input('Are you sure you want to set :\t{}\t=\t{}\t?'.format(variable_name, value))
+        if proceed in ['n', 'N']:
+            continue
+        elif proceed in ['y', 'Y']:
+            return value
+            break
+        else:
+            print('Please Enter y/n.')
+
```
Comment (paulboone) [inline htsohm/htsohm.py]: Try something like this:

def are_you_sure()
    while True:
        proceed = input('Are you sure? [y/n])
        if proceed in ['n', 'N']:
            return False
        elif proceed in ['y', 'Y']:
            return True        
        else:
            print('Please Enter y/n.')

def set_variable_better_name(message, variable_name):
    while True:
        value = input(message)
        print("Set var ### to ###")
        if are_you_sure():
            return value

**Review by paulboone — CHANGES_REQUESTED** [review]
- refactor are_you_sure
- comments for select_parent blocks
- Session => session
- add comment to say why this is here:

    if config['interactive_mode'] == 'on':
        p_x, p_y, p_z = [], [], []
        for atom_site in child_material.structure.atom_sites:
            p_x.append(atom_site.x_frac)
            p_y.append(atom_site.y_frac)
            p_z.append(atom_site.z_frac)

For later:
- config refactor

**Review by paulboone — APPROVED** [review]
